src/preprocess/etl.py

```diff
@@ -0,0 +1,227 @@
+
+        # select the most recent data
+        file_path = files.sort_values(time_col_name, ascending=False)[parent_path_col_name][0].joinpath(
+            files.sort_values(time_col_name, ascending=False)[file_col_name][0])
+
+        # read in raw data
+        df = pd.read_csv(file_path)
+
+        # drop cols without headers
+        df = df[df.columns.dropna()]
+
+        # import type.yml to specify column types
+        yml_file = pl.Path(__file__).resolve().parents[0].joinpath('type.yml')
+        with open(yml_file) as f:
+            # use safe_load instead load
+            type_dic = yaml.safe_load(f)
+        col_with_ws = type_dic['Trim_ws']
+
+        # convert white space to nan
+        df[col_with_ws] = df[col_with_ws].replace(r'^\s*$', np.nan, regex=True)
+
+        df.to_csv(file_path, index=False)
+        print("data conversion done!")
+
+        return
+
+    @staticmethod
+    def load(df: pd.DataFrame, file_col_name: str = "File_Name", time_col_name: str = "Created",
+             parent_path_col_name: str = "Parent", contains_str: str = "datasets", yml_file: str = 'type.yml',
+             name_in_reference: str = 'churn'):
+
+        """
+        Find the most recent data file and load the data
+        Args:
+            df: pandas df with all the files in the directory
+            file_col_name: column name of file names in df
+            time_col_name: column name of created time in df
+            parent_path_col_name: column name of parent path in df
+            contains_str: data input file should have this str in the file name
+            name_in_reference: the key in type.yml for the corresponding data set
+        Returns: pandas df of input data
+        """
+
+        # select all the files containing name "datasets"
+        files = df[df[file_col_name].str.contains(contains_str, na=False)].reset_index()
+        if len(files.index) >= 1:
+            # select the most recent data
+            file_path = files.sort_values(time_col_name, ascending=False)[parent_path_col_name][0].joinpath(
+                files.sort_values(time_col_name, ascending=False)[file_col_name][0])
+
+            # import type.yml to specify column types
+            yml_file = pl.Path(__file__).resolve().parents[0].joinpath(yml_file)
+            with open(yml_file) as f:
+                # use safe_load instead load
+                type_dic = yaml.safe_load(f)
+
+            type_reference = type_dic[name_in_reference]
+
+            # load data
+            data = pd.read_csv(file_path, header=0, dtype=type_reference)
+
+            # lower case for column names
+            data.columns = map(str.lower, data.columns)
+
+            return data
+        else:
+            raise ValueError("file not exists!")
+
+    @staticmethod
+    def check_size(data: pd.DataFrame, n_cols: int = 21):
+        """
+        Check # of columns and if the data is empty
```

Comment:
if empty, do what?

Reply:
good point to separate empty vs. other value errors, added an "if" condition for empty data